parquet-hadoop/src/main/java/org/apache/parquet/hadoop/MemoryManager.java

```diff
@@ -74,7 +75,7 @@ private void checkRatio(float ratio) {
    * @param writer     the new created writer
    * @param allocation the requested buffer size
    */
-  synchronized void addWriter(InternalParquetRecordWriter<?> writer, Long allocation) {
+  void addWriter(InternalParquetRecordWriter<?> writer, Long allocation) {
```

Comment:
This is not the same. The `synchronized` locks the whole method. While the `ConcurrentHashMap` could only guarantee the safety update here.

Reply:
@ConeyLiu thank you for a prompt review. Excited to get this fixed. Agree it's not the same, and further `synchronized` is not locking a whole method BUT a whole **instance** which in the case of MemoryManager is the **single/global instance**, so `synchronized` can be very expensive here. The cost is manifested by the lock stats above. I can see the tests are all green. Is such an "excessive" synchronization pattern on the global instance necessary for the semantic of MemoryManager? Which use-cases require this level of locking, and could that be relaxed? What do you recommend as next steps? Thanks in advance!

Reply:
👋 @ConeyLiu I hope you're doing well. I wanted to gently follow up on this thread. Your input would be really valuable for moving forward. Thank you.

Reply:
I agree with @ConeyLiu that there is a behavior change. `updateAllocation()` is not thread-safe and may have race condition after this change. I suspect the unit test does not cover the case where multiple writers are added and deleted concurrently.

Reply:
Unfortunately I do not have too much experience in using this code path. The MemoryManager was introduced long ago to address memory issue for dynamic partition. I'm not sure if Apache Spark or other engines are using it or not. May I seek advise from @dongjoon-hyun @sunchao @parthchandra @shangxinli to see if you have seen this problem before?

Reply:
@wgtmac` I'm not too familiar with this part of the code path either. However, it does look that the `updateAllocation` call needs to be synchronized and it requires the `writers` map to not be mutated during that time. 
Spark uses the default behaviour of ParquetOutputFormat and so does end up using the MemoryManager. The impact of this change is not clear to me though.
BTW, @ravwojdyla this is a great bit of investigation you did there! Would it be possible to share the jfr file (in the jira, perhaps)? I'd like to see if I can gain any additional insights from the information you have collected.

Reply:
@parthchandra thanks 🙇I unfortunately don't have those specific files anymore, it's been a couple of months actually (June 2023), apologies. If this PR continues, I may need to run another profiling to validation the patch, I would be happy to share the before and after jfr files.

Reply:
NP. Can I suggest a unit test that adds/removes writers concurrently? Almost sure that the test is likely to fail because of `updateAllocation`. The question remains whether such concurrent access does, in fact, occur in engines like Spark or Hadoop.

Reply:
Does this have a big impact on the end-to-end performance? I think it should only happen when you are writing a file with many small partitions.

Reply:
@ConeyLiu yes it does. One of our production Spark job took close to 2 hours with partitioned parquet output and about 45 minutes without. Also please see the stats in JIRA issues:
 * https://issues.apache.org/jira/browse/PARQUET-2412
 * https://issues.apache.org/jira/browse/SPARK-44003

Reply:
Thanks @ravwojdyla for your sharing. I think the changes should keep the safety. It needs more UTs to verify this.